lib/optimizer/src/Transformer/ServerSideRendering.php

```diff
@@ -620,4 +699,271 @@ private function isAmpExperimentUsed(DOMElement $element)
+     * @param DOMElement $element        Element to adapt.
+     * @param string     $attributeName  Name of the attribute to be extracted.
+     * @param string     $attributeValue Value of the attribute to extract the CSS styling from.
+     * @return string Extract custom CSS styling.
+     */
+    private function extractSizesAttributeCss(Document $document, DOMElement $element, $attributeName, $attributeValue)
+    {
+        if (!$element->hasAttribute(Attribute::SRCSET) || empty($element->getAttribute(Attribute::SRCSET))) {
+            // According to the Mozilla docs, a sizes attribute without a valid srcset attribute should have no effect.
+            // Therefore, it should simply be stripped, without producing media queries.
+            // @see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attr-sizes
+            return '';
+        }
+
+        return $this->extractAttributeCss(
+            $document,
+            $element,
+            $attributeName,
+            $attributeValue,
+            '#__ID__{width:%s}',
+            '@media %s{#__ID__{width:%s}}'
+        );
+    }
+
+    /**
+     * Extract the custom CSS styling from a 'heights' attribute.
+     *
+     * __heights__ (AMP-specific)
+     * "The value of this attribute is a sizes expression based on media expressions, similar to the sizes attribute on
+     * img tags but with two key differences:
+     * - The value applies to the height, not the width of the element.
+     * - Percent values are allowed. A percent value indicates the percent of the element's width. For example, a value
+     *   of 80% indicates that the height of the element will be 80% of the element's width."
+     *
+     * @see https://amp.dev/documentation/guides-and-tutorials/learn/common_attributes/#heights
+     *
+     * @param Document   $document       Document containing the element to adapt.
+     * @param DOMElement $element        Element to adapt.
+     * @param string     $attributeName  Name of the attribute to be extracted.
+     * @param string     $attributeValue Value of the attribute to extract the CSS styling from.
+     * @return string Extract custom CSS styling.
+     */
+    private function extractHeightsAttributeCss(Document $document, DOMElement $element, $attributeName, $attributeValue)
+    {
+        /*
+         * Note: Usage of :first-child could potentially be brittle here.
+         * We cannot use a selector that references the .i-amphtml-sizer element directly because that reference doesn't
+         * pass the validator.
+         * See https://github.com/ampproject/amp-wp/pull/4482#discussion_r421149537
+         */
+        return $this->extractAttributeCss(
+            $document,
+            $element,
+            $attributeName,
+            $attributeValue,
+            '#__ID__:first-child{height:%s}',
+            '@media %s{#__ID__:first-child{height:%s}}'
```

Comment:
This won't work as expected, as the `:first-child` is a pseudo-selector qualifying what it is attached to. So this would match elements that have ID of `__ID__` which are the first-child position. So it needs rather to be:

```suggestion
            '#__ID__>:first-child{height:%s}',
            '@media %s{#__ID__>:first-child{height:%s}}'
```

So it selects the first child of the element that has the ID.

Reply:
Ah, you are right. Will correct.

Reply:
... I just realized that it is not the `height` that needs to be adapted for the sizer, I think, but the `padding-top`.

Reply:
... which can't be set as it is inlined by JS already. 🤔

Reply:
So, as the Optimizer is the one that inlined the `padding-top`, I could skip that if I detect that I already had a `heights` attribute that turned into CSS.

But at what point are we just plain "hacking" here?

Reply:
Oh, I could actually store the CSS and properly inline it onto the sizer. That way, I can omit adding it to the size-limited `<style amp-custom>` as well.

Reply:
I'm running in circles here. Inline styles don't allow for media queries. I will just skip adding the `padding-top` in that case for now.

Reply:
Ok, I've got a working version now. Not the most elegant stuff, but it seems to work as expected.